src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt

```diff
@@ -202,75 +44,64 @@ class ElectrumClient(
-            _connectionState.filter { it != previousState }.collect {
+            fun convert(input: ElectrumConnectionStatus): Connection = when (input) {
+                is ElectrumConnectionStatus.Connecting -> Connection.ESTABLISHING
+                is ElectrumConnectionStatus.Connected -> Connection.ESTABLISHED
+                is ElectrumConnectionStatus.Closed -> Connection.CLOSED(input.reason)
+            }
+
+            var previousState = convert(_connectionStatus.value)
+            _connectionStatus.map { convert(it) }.filter { it != previousState }.collect {
                 logger.info { "connection state changed: ${it::class.simpleName}" }
-                if (it is Connection.CLOSED) mailbox.send(Disconnected)
+                _connectionState.value = it
                 previousState = it
             }
         }
-        runJob = launch { run() }
     }
 
-    private suspend fun run() {
-        mailbox.consumeEach { cmd ->
-
-            logger.debug { "processing command $cmd in state ${state::class.simpleName}" }
-
-            val (newState, actions) = state.process(cmd, logger)
-            state = newState
-
-            actions.forEach { action ->
-                yield()
-                when (action) {
-                    is SendRequest -> if (!output.isClosedForSend) output.send(action.request.encodeToByteArray())
-                    is SendHeader -> _notifications.emit(ElectrumClientNotification(action.callerId, HeaderSubscriptionResponse(action.height, action.blockHeader)))
-                    is SendResponse -> _notifications.emit(ElectrumClientNotification(action.callerId, action.response))
-                    is BroadcastStatus -> if (_connectionState.value != action.connection) _connectionState.value = action.connection
-                }
-            }
-        }
-    }
 
     fun connect(serverAddress: ServerAddress) {
-        if (_connectionState.value is Connection.CLOSED) establishConnection(serverAddress)
-        else logger.warning { "electrum client is already running" }
+        if (_connectionStatus.value is ElectrumConnectionStatus.Closed) {
+            runJob = establishConnection(serverAddress)
+        } else logger.warning { "electrum client is already running" }
     }
 
     fun disconnect() {
-        if (this::socket.isInitialized) socket.close()
-        _connectionState.value = Connection.CLOSED(null)
-        output.close()
+        launch {
+            if (!mailbox.isClosedForSend) {
+                mailbox.send(Action.Disconnect)
+            }
+        }
     }
 
-    // Warning : lateinit vars have to be used AFTER their init to avoid any crashes
-    //
-    // This shouldn't be used outside the establishedConnection() function
-    // Except from the disconnect() one that check if the lateinit var has been initialized
-    private lateinit var socket: TcpSocket
     private fun establishConnection(serverAddress: ServerAddress) = launch {
```

Comment:
On Android with Tor enabled and a flaky connection, I've been able to sometimes crash the app because of `linesFlow` below. `linesFlow` relies on `JvmSocket.receiveAvailable` (on Android) which may throw `TcpSocket.IOException.ConnectionClosed`. This exception is not caught and will crash the `establishConnection` coroutine (and the app).

This is probably a race condition that Tor makes more obvious. I don't have a good solution for this for now and this is probably outside the scope of this PR. But in any case, I think we should add a `CoroutineExceptionHandler` here that will catch errors happening during initialisation, log the error and make sure the socket is closed.

Reply:
done in https://github.com/ACINQ/lightning-kmp/pull/411/commits/f66fd8221e492a36567df03027935e8879691679